listenbrainz/webserver/static/js/src/fresh-releases/ReleaseTimeline.tsx

```diff
@@ -0,0 +1,87 @@
+import * as React from "react";
+import Slider from "rc-slider";
+import { countBy, debounce, zipObject } from "lodash";
+import { formattedReleaseDate, useMediaQuery } from "./utils";
+
+type ReleaseTimelineProps = {
+  releases: Array<FreshReleaseItem>;
+};
+
+export default function ReleaseTimeline(props: ReleaseTimelineProps) {
+  const { releases } = props;
+
+  const [currentValue, setCurrentValue] = React.useState<number | number[]>();
+  const [marks, setMarks] = React.useState<{ [key: number]: string }>({});
+
+  const screenMd = useMediaQuery("(max-width: 992px)"); // @screen-md
+
+  const changeHandler = React.useCallback((percent: number | number[]) => {
+    setCurrentValue(percent);
+    const element: HTMLElement | null = document.getElementById(
+      "release-cards-grid"
+    )!;
+    const scrollHeight = ((percent as number) / 100) * element.scrollHeight;
+    const scrollTo = scrollHeight + element.offsetTop;
+    window.scrollTo({ top: scrollTo, behavior: "smooth" });
+    return scrollTo;
+  }, []);
+
+  function createMarks(data: Array<FreshReleaseItem>) {
+    const releasesPerDate = countBy(
+      releases,
+      (item: FreshReleaseItem) => item.release_date
+    );
+    const datesArr = Object.keys(releasesPerDate).map((item) =>
+      formattedReleaseDate(item)
+    );
+    const percentArr = Object.values(releasesPerDate)
+      .map((item) => (item / data.length) * 100)
+      .map((_, index, arr) =>
+        arr.slice(0, index + 1).reduce((prev, curr) => prev + curr)
+      );
+    percentArr.unshift(0);
+    percentArr.pop();
+    const middle = percentArr[Math.floor(percentArr.length / 2)];
+
+    // Scroll to the current date
+    setCurrentValue(middle);
+    window.scrollTo({ top: changeHandler(middle), behavior: "smooth" });
+
+    return zipObject(percentArr, datesArr);
+  }
+
+  const handleScroll = React.useCallback(
+    debounce(() => {
+      // TODO change to relative position of #release-cards-grid instead of window
```

Comment:
Todo todo?
(todo todo-todo-todo-todo-todooooooo  todododo)
https://images.squarespace-cdn.com/content/v1/566a4af357eb8d3974390587/1454570174741-4Y47MU9KC0IEMF87J7JE/ke17ZwdGBToddI8pDm48kHnqq_UrsdjXT8IYn8eSSjYUqsxRUqqbr1mOJYKfIPR7LoDQ9mXPOjoJoqy81S2I8PaoYXhp6HxIwZIk7-Mi3Tsic-L2IOPH3Dwrhl-Ne3Z2qO_jwV24K38aH_dPXd7Kl7ikppHhGne0pYe9e5QogUDdZR9z9mxWb0yLUToVqwSd/PinkPantherPinkSax-LtdEdCel.jpg

Reply:
Hahaha 🤣 

It's for the scroll logic problem that I mentioned in the description.

Reply:
Logic behind the `unshift(0)` and `pop()` - 
It may look hacky.
And yes, that's the reason there is no date at the end of the slider but there is one at the start (because of `unshift(0)`).

If we don't do that, the timeline ends up looking like this - 
![image](https://user-images.githubusercontent.com/6230769/199591758-3731eacd-2ac4-46d0-b518-9fb779a39ef8.png)

We want the timeline dates or marks to start where the grid starts. So the 0% should always have the first date. With the same logic, we don't want the last date to be at 100% because that will mean we're at the bottom of the grid. The last date should start before 100%. That explains the `pop()`.

Reply:
Yeah, I see it now, thanks for the explanation. A comment would be useful for posterity.

Reply:
Done